src/VirtualDevice.ts

```diff
@@ -305,8 +308,106 @@ export class VirtualDevice {
+        const mesageProcessor = new MessageProcesor(messages);
+        return await mesageProcessor.process();
+    }
 }
 
+class MessageProcesor {
+    public constructor( public messages: IMessage[]) {}
+
+    public process(): Promise<IMessageEndpoint[]> {
+        return Promise.all(this.messages.map((message) => this.processMessage(message)));
+    }
+
+    private async processMessage(message: IMessage): Promise<IMessageEndpoint> {
+        let fileContent: string | undefined;
+        let extension: string | undefined;
+
+        const frameRate: number | undefined =  message.audio ? message.audio.frameRate : undefined;
+        const channels: number | undefined =  message.audio ? message.audio.channels : undefined;
+        const sampleWidth: number | undefined =  message.audio ? message.audio.sampleWidth : undefined;
+
+        let filePath: string;
+
+        if (message.audio) {
+            if (message.audio.audioPath || message.audio.audioURL) {
+                if (message.audio.audioPath) {
+                    filePath = message.audio.audioPath;
+                    fileContent = this.getLocalFile(message.audio.audioPath);
+                } else if (message.audio.audioURL) {
+                    filePath = message.audio.audioURL;
+                    fileContent = await this.fetchFile(message.audio.audioURL);
+                } else {
+                    filePath = "";
+                    fileContent = "";
+                }
+
+                // get extension
+                extension =  this.getExtension(filePath);
+
+            } else {
+                throw new Error("Invalid audio");
+            }
+        }
+
+        return {
+            audio: fileContent,
+            channels,
+            format: extension,
+            frame_rate: frameRate,
+            phrases: message.phrases,
+            sample_width: sampleWidth,
+            text: message.text,
+        } as IMessageEndpoint;
+    }
+
+    private getLocalFile(path: string) {
+        const fileContents = fs.readFileSync(path);
+        return Buffer.from(fileContents).toString("base64");
+    }
+
+    private fetchFile(url: string): Promise<string> {
+        return new Promise((resolve, reject) => {
+            const data: Buffer[] = [];
+            const req = https.get(url as any, (res) => {
+                res.on("data", (chunk: Buffer) => {
+                    data.push(chunk);
+                });
+                res.on("end", () => {
+                    if (res.statusCode === 200) {
+                        const buffer = Buffer.concat(data);
+                        resolve(buffer.toString("base64"));
+                    } else {
+                        reject();
+                    }
+                });
+            });
+            req.on("error", function(error: Error) {
+                reject(error.message);
+            });
+            req.end();
+        });
+    }
+
+    private getExtension(path: string) {
```

Comment:
you can use node path library

import * as path from 'path';
path.extname('index.html');